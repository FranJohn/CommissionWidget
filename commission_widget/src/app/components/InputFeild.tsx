import React, { ChangeEvent } from 'react';

interface InputFieldProps {
  placeholder?: string;
  onChange: (value: number) => void;
}

/**
 * InputField component for rendering an input field for commission value entry.
 * @param {InputFieldProps} props - Properties for the InputField component.
 * @returns {JSX.Element} - JSX for the InputField component.
 */
const InputField: React.FC<InputFieldProps> = ({ placeholder = '', onChange }) => {
    /**
     * Function to handle changes in the input field.
     * @param {ChangeEvent<HTMLInputElement>} event - The change event object.
     * @returns {void}
     */
    const handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
        const { value } = event.target;
        // Ensure the input can only be numbers
        const sanitizedValue = value.replace(/[^\d.]|(?<=\..*)\./g, '');
        if (value !== sanitizedValue) {
            // Show an alert if non-numeric characters are entered
            alert('Please enter only numbers.');
        }
        onChange(+sanitizedValue);
    };

    /**
     * Function to handle focus on the input field.
     * Removes incorrect characters while keeping valid characters intact.
     * @param {React.FocusEvent<HTMLInputElement>} event - The focus event object.
     * @returns {void}
     */
    const handleInputFocus = (event: React.FocusEvent<HTMLInputElement>) => {
        const { value } = event.target;
        const newValue = value.replace(/[^\d.]|(?<=\..*)\./g, ''); // Remove only wrong characters
        if (value !== newValue) {
            event.target.value = newValue;
            onChange(+newValue);
        }
    };


    // Input type is text to allow decimal points for fractions (type number not always supported in old browsers)
    return (
        <div className="input-wrapper">
            <input
                type="text"
                placeholder={placeholder}
                onChange={handleInputChange}
                onFocus={handleInputFocus}
                className="custom-input"
            />
        </div>
    );
};

export default InputField;
