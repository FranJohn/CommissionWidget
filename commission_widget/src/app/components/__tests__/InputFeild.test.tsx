import '@testing-library/jest-dom';
import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import InputFeild from '../InputFeild';

jest.spyOn(window, 'alert').mockImplementation(() => {});

describe('InputFeild component', () => {
    it('should render with placeholder text', () => {
        const { getByPlaceholderText } = render(<InputFeild placeholder="Enter value" onChange={() => {}} />);
        const inputElement = getByPlaceholderText('Enter value');
        expect(inputElement).toBeInTheDocument();
    });

    it('should handle change events', () => {
        const onChange = jest.fn();
        const { getByPlaceholderText } = render(<InputFeild placeholder="Enter value" onChange={onChange} />);
        const inputElement = getByPlaceholderText('Enter value');
        
        fireEvent.change(inputElement, { target: { value: '1234567890' } });
        expect(onChange).toHaveBeenCalledWith(1234567890);

        fireEvent.change(inputElement, { target: { value: 't1&f4!n3?' } });
        expect(onChange).toHaveBeenCalledWith(143);

        fireEvent.change(inputElement, { target: { value: '456.789' } });
        expect(onChange).toHaveBeenCalledWith(456.789);

        fireEvent.change(inputElement, { target: { value: '956..' } });
        expect(onChange).toHaveBeenCalledWith(956);

        fireEvent.change(inputElement, { target: { value: '9.5.6.' } });
        expect(onChange).toHaveBeenCalledWith(9.56);
    });

    it('should handle focus events', () => {
        const onChange = jest.fn();
        const { getByPlaceholderText } = render(<InputFeild placeholder="Enter value" onChange={onChange} />);
        const inputElement = getByPlaceholderText('Enter value');

        fireEvent.focus(inputElement);
        expect(onChange).not.toHaveBeenCalled();

        fireEvent.change(inputElement, { target: { value: '34' } });
        fireEvent.focus(inputElement);
        expect(onChange).toHaveBeenCalledWith(34);
    });

    it('matches the InputFeild snapshot', () => {
        const onChange = jest.fn();
        const { container } = render(<InputFeild placeholder="Enter value" onChange={onChange} />);
        expect(container).toMatchSnapshot();
    });
});
