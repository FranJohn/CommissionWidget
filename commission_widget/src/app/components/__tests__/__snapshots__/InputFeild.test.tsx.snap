// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputFeild component matches the InputFeild snapshot 1`] = `
<div>
  <div
    class="input-wrapper"
  >
    <input
      class="custom-input"
      placeholder="Enter value"
      type="text"
    />
    GBP
  </div>
</div>
`;
